darts/models/forecasting/query_selector.py

```diff
@@ -0,0 +1,650 @@
+
+        self.enc_dropout = MonteCarloDropout(p=dropout)
+
+        # Dense layers for managing network inputs and outputs
+        self.enc_input_fc = nn.Linear(self.input_dim, dim_val)
+        self.dec_input_fc = nn.Linear(self.input_dim, dim_val)
+        self.out_fc = nn.Linear(
+            dec_seq_len * dim_val,
+            self.output_chunk_length * self.output_dim * self.nr_params,
+        )
+
+        self.debug = debug
+
+    def forward(self, x_in: tuple):
+        x, _ = x_in
+        # x: [Batch, Input length, Channel]
+
+        # encoder
+        e = self.encs[0](self.pos(self.enc_dropout(self.enc_input_fc(x))))
+
+        for enc in self.encs[1:]:
+            e = enc(e)
+        if self.debug:
+            print(f"Encoder output size: {e.shape}")
+        # decoder
+        decoded = self.dec_input_fc(x[:, -self.dec_seq_len :])
+
+        d = self.decs[0](decoded, e)
+        for dec in self.decs[1:]:
+            d = dec(d, e)
+
+        # output
+        x = self.out_fc(d.flatten(start_dim=1))
+
+        x = torch.reshape(x, (x.shape[0], -1, self.output_dim))
+        x = x.view(
+            x.shape[0], self.output_chunk_length, self.output_dim, self.nr_params
+        )
+        return x
+
+    def record(self):
+        self.debug = True
+        for enc in self.encs:
+            enc.record()
+        for dec in self.decs:
+            dec.record()
+
+
+class QuerySelectorModel(PastCovariatesTorchModel):
+    def __init__(
+        self,
+        input_chunk_length: int,
+        output_chunk_length: int,
+        dim_val: int = 32,
+        dim_attn: int = 128,
+        dec_seq_len: int = 12,
+        n_encoder_layers: int = 3,
+        n_decoder_layers: int = 3,
+        enc_attn_type="query_selector_0.8",
+        dec_attn_type="full",
+        n_heads: int = 4,
+        dropout: float = 0.1,
+        feed_forward: str = "default",
+        norm_type: str = "LayerNorm",
+        **kwargs,
+    ):
+        """An implementation of the QuerySelector model, as presented in [1]_.
+
+
+
+        Parameters
+        ----------
+        input_chunk_length : int
+            The length of the input sequence fed to the model.
+        output_chunk_length : int
+            The length of the forecast of the model.
+        dim_val : int
+            hidden dimension size of the value in the attention layer (default=32).
+        dim_attn : int
+            hidden dimension size of the attention layer (default=128).
+        dec_seq_len : int
+            the embedding size for the decoder (default=12).
+        n_encoder_layers : int
+            The number of encoder layers in the encoder (default=3).
+        n_decoder_layers : int
+            The number of decoder layers in the encoder (default=3).
+        enc_attn_type : str
+            "full" attention or query selector with a reduction f factor. Example "query_selector_0.4".
+            Default to "query_selector_0.8"
+        dec_attn_type : str
```

Comment:
Could we refactor our current `Transformer` model to somehow leverage this and only be a particular case of query selector with full matrix?

Reply:
Could we also by any chance incorporate `Informer` as another particular implementation here?